vmm-reference-gdt.md

```diff
@@ -0,0 +1,185 @@
+    fn segment_type(&self) -> u8;
+```
+`create_kvm_segment` creates a `kvm_segment` from the `SegmentDescriptor`. 
+
+`kvm_segment` is the data type of all the segment registers (`cs`, `ds`, etc.) in a KVM vCPU.
+
+`table_index` is the index of the entry in the gdt table.
+```rs
+fn create_kvm_segment(&self, table_index: usize) -> kvm_segment {
+        kvm_segment {
+            base: self.base(),
+            limit: self.limit(),
+            // The multiplication is safe because the table_index can be maximum
+            // `MAX_GDT_SIZE`. The conversion is safe because the result fits in u16.
+            selector: (table_index * 8) as u16,
+            type_: self.segment_type(),
+            present: self.p(),
+            dpl: self.dpl(),
+            db: self.db(),
+            s: self.s(),
+            l: self.l(),
+            g: self.g(),
+            avl: self.avl(),
+            padding: 0,
+```
+The `p` bit in flags specifies whether the segment is valid or not. `p` is set to 1 for a valid segment.
+``` rs
+            unusable: match self.p() {
+                0 => 1,
+                _ => 0,
+            },
+        }
+    }
+```
+
+
+
+## GDT
+Implemented as a vector of segment descriptors.
+```rs
+pub struct Gdt(Vec<SegmentDescriptor>);
+```
+`try_push` tries to push an entry into the `GDT`. An error is thron if the number of entries is greater than $2^{13} - 1$   (max 12 bits).
+```rs
+pub fn try_push(&mut self, entry: SegmentDescriptor) -> Result<()>
+```
+
+`create_kvm_segment_for` returns a `kvm_segment` corresponding to the GDT entry at `index`.
+```rs
+pub fn create_kvm_segment_for(&self, index: usize) -> Option<kvm_segment>
+```
+
+This method writes the GDT into Guest Memory. 
+For each entry, the address is calculated first. The base of the GDT is at `boot_gdt_addr`. Each entry is placed at the end of the previous entry.
+```rs
+pub fn write_to_mem<Memory: GuestMemory>(&self, mem: &Memory) -> Result<()>;
+```
+
+A default implementation of the GDT with default segment descriptors has been provided at the line 257. In the default table, for all the segments, base is 0 and limit is $2^{20}$. 
+```rs
+impl Default for Gdt {
+    fn default() -> Self {
+        Gdt(vec![
+            SegmentDescriptor::from(0, 0, 0),            // NULL
+            SegmentDescriptor::from(0xa09b, 0, 0xfffff), // CODE
+            SegmentDescriptor::from(0xc093, 0, 0xfffff), // DATA
+            SegmentDescriptor::from(0x808b, 0, 0xfffff), // TSS - holds information about the task's state
+        ])
+    }
+}
+```
+
+
+## IDT
+This module also has a function to set a value at the IDT offset. It has only been used once, to create an empty IDT entry.
+```rs
+pub fn write_idt_value<Memory: GuestMemory>(val: u64, guest_mem: &Memory) -> Result<()>;
+```
+
+
+<br /><br /><br />
+
+This module has been used in `KvmVcpu` in the `configure_sregs` method to set the base and limit of the GDT, the IDT and to setup the segment registers.
```

Comment:
would be good to link to the file in https://github.com/codenet/understanding-rust-vmm/ that talks about `configure_sregs`